### PR TITLE
Issue #1352 - Show saved nameservers if registry is down

### LINKED ISSUE #1352: Show saved nameservers if registry is down
### Story

As a registrant
I want to see my nameservers even if the registry is down
so that I am not worried about missing nameservers and so that I don't see the "add nameservers" button on the overview page

### Acceptance Criteria

- [ ] nameservers and their ip address are saved in our database when received from epp
- [ ] anytime nameservers/ips is recieved from epp, update our database object accordingly including removing them
- [ ] if epp is down and a nameserver is saved for that domain, that is shown instead of nothing  (this should be done on the overview page and nameserver page)

### Additional Context

note we already have a nameserver and host_ip tables that could be used, please evaluate them and make any changes as needed.

### Issue Links

Relates to #1336 (Consider this in design decisions for handling this condition)

### PR DESCRIPTION
**Pull request by dave-kennedy-ecs**
## Ticket

Resolves #1352 

## Changes

- DNS Nameservers (host names and IPs) are stored in the registrar database when they are fetched from registry through EPP
- When registry is unavailable through EPP, DNS Nameservers are retrieved from the registrar database rather than from the registry through EPP

#### Satisfied acceptance criteria and met development standards

- [ ] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [ ] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/registrar/models/domain.py [hunk 1]
```diff
@@ -1666,6 +1670,51 @@ def _update_hosts_and_contacts(self, cleaned, fetch_hosts, fetch_contacts):
...
+            cleaned: dict containing hosts.  Hosts are provided as a list of dicts, e.g.
+                [{"name": "ns1.example.com",}, {"name": "ns1.example.gov"}, "addrs": ["0.0.0.0"])]
+            fetch_hosts: boolean indicating whether or not fetch_hosts was called
+        """
+        if fetch_hosts:
+            cleaned_hosts = cleaned["hosts"]
+            # Get all existing hosts from the database for this domain
+            existing_hosts_in_db = Host.objects.filter(domain=self)
+            # Identify hosts to delete
+            cleaned_host_names = set(cleaned_host["name"] for cleaned_host in cleaned_hosts)
```
Comment (therealslimhsiehdy) [inline src/registrar/models/domain.py:1687]: **Question:** Would you mind giving an example of when we would have duplicates of a host name? (Just trying to think through all edge cases)

**Review by therealslimhsiehdy — COMMENTED** [review]
**On local:** Bc EPP doesn't work on local I was able to test out the scenario of retrieving from the registrar DB (via Django Admin > Host > Add new host) rather than from the registry through EPP
Screenshot 2023-12-28 at 4 02 19 PM

**On sandbox:** When I add in Host(s) via Django Admin and then make changes via the UI to these Hosts + IPs, I see the changes reflected in the Django Admin -- this means that the database is being updated alongside
Screenshot 2023-12-28 at 4 35 44 PM
Screenshot 2023-12-28 at 4 35 54 PM

**Question:** Still milling about the edge case of adding 1 in via Hosts in Django Admin and another via UI and it having the section bork where I can't even add in new Nameservers and IPs even when I deleted it from Hosts. 

Even if we set it it as readonly on Django Admin, wouldn't we still need to apply certain permission settings to add in Hosts (ie make sure they put 2 Nameservers in for 1 domain, but I feel like this may be tricky) to safeguard against unwanted data errors even for an Admin permission level user?

**Comment by rachidatecs** [comment]
The doctstrings for the Host and HostIp classes need updating. They still say: "This model exists ONLY to allow a new registrant to draft DNS entries before their application is approved."

File: src/registrar/models/domain.py (hunk 1 above)
Comment (dave-kennedy-ecs) [reply src/registrar/models/domain.py:1687]: There should not be a case of duplicates.  It is not allowed by the logic in the name server form; throws an error 'Remove duplicate entry'.  And it is not allowed by the registry.

**Comment by dave-kennedy-ecs** [comment]
> The doctstrings for the Host and HostIp classes need updating. They still say: "This model exists ONLY to allow a new registrant to draft DNS entries before their application is approved."

Updated the docstrings

**Comment by dave-kennedy-ecs** [comment]
> **On local:** Bc EPP doesn't work on local I was able to test out the scenario of retrieving from the registrar DB (via Django Admin > Host > Add new host) rather than from the registry through EPP Screenshot 2023-12-28 at 4 02 19 PM
> 
> **On sandbox:** When I add in Host(s) via Django Admin and then make changes via the UI to these Hosts + IPs, I see the changes reflected in the Django Admin -- this means that the database is being updated alongside Screenshot 2023-12-28 at 4 35 44 PM Screenshot 2023-12-28 at 4 35 54 PM
> 
> **Question:** Still milling about the edge case of adding 1 in via Hosts in Django Admin and another via UI and it having the section bork where I can't even add in new Nameservers and IPs even when I deleted it from Hosts.
> 
> Even if we set it it as readonly on Django Admin, wouldn't we still need to apply certain permission settings to add in Hosts (ie make sure they put 2 Nameservers in for 1 domain, but I feel like this may be tricky) to safeguard against unwanted data errors even for an Admin permission level user?

I have removed Hosts from django admin.  For two reasons:
1 - any changes made in django admin will never be propagated to registry, so allowing admins to update this data in django admin would be misleading
2 - the logic which enforces rules about valid name servers is not applied in django admin and allows for data which can break functionality for domain management in the app

File: src/registrar/models/domain.py [hunk 2]
```diff
@@ -1666,6 +1670,51 @@ def _update_hosts_and_contacts(self, cleaned, fetch_hosts, fetch_contacts):
...
+    def _update_hosts_and_ips_in_db(self, cleaned, fetch_hosts):
+        """Update hosts and host_ips in database if retrieved from registry.
+
+        Parameters:
+            self: the domain to be updated with hosts and ips from cleaned
+            cleaned: dict containing hosts.  Hosts are provided as a list of dicts, e.g.
+                [{"name": "ns1.example.com",}, {"name": "ns1.example.gov"}, "addrs": ["0.0.0.0"])]
+            fetch_hosts: boolean indicating whether or not fetch_hosts was called
+        """
+        if fetch_hosts:
```
Comment (rachidatecs) [inline src/registrar/models/domain.py:1682]: What do you think about putting this guy outside the method?

File: src/registrar/models/domain.py [hunk 3]
```diff
@@ -1651,7 +1655,7 @@ def _get_dnssec_data(self, response_extensions):
         return dnssec_data
 
     def _update_hosts_and_contacts(self, cleaned, fetch_hosts, fetch_contacts):
-        """Capture and store old hosts and contacts from cache if they don't exist"""
+        """Capture and cache old hosts and contacts from cache if they don't exist in cleaned"""
```
Comment (rachidatecs) [inline src/registrar/models/domain.py:1658]: """Update cache from old cache"""

Is this what the method's doing?

**Review by rachidatecs — APPROVED** [review]

File: src/registrar/models/domain.py (hunk 2 above)
Comment (therealslimhsiehdy) [reply src/registrar/models/domain.py:1682]: +1 on this, but definitely non-blocking!

**Review by therealslimhsiehdy — APPROVED** [review]

File: src/registrar/models/domain.py (hunk 3 above)
Comment (dave-kennedy-ecs) [reply src/registrar/models/domain.py:1658]: You're right, I am updating the comment before merging

File: src/registrar/models/domain.py (hunk 2 above)
Comment (dave-kennedy-ecs) [reply src/registrar/models/domain.py:1682]: I agree.  Done
